credoai/lens/lens_validation.py

```diff
@@ -0,0 +1,172 @@
+    whether the model supports that function. If so, the validator applies the outputting function to
+    a small sample of the supplied dataset. The validator ensures the outputting function does not fail
+    and, depending on the nature of the outputting function, performs light checking to verify the outputs
+    (e.g. predictions) match the expected form, possibly including: data type and output shape.
+
+    Parameters
+    ----------
+    model : artifacts.Model or a subtype of artifacts.Model
+        A trained machine learning model wrapped as a Lens Model object
+    data : artifacts.Data or a subtype of artifacts.Data
+        The dataset that will be assessed by Lens evaluators, wrapped as a Lens Data object
+    """
+    # check predict
+    # Keras always outputs numpy types (not Tensor or something else)
+    if "predict" in model.__dict__.keys():
+        try:
+            mini_pred, batch_size = check_prediction_model_output(model.predict, data)
+            if not mini_pred.size:
+                # results for all presently supported models are ndarray results
+                raise Exception("Empty return results from predict function.")
+            if isinstance(data.y, np.ndarray) and (
+                mini_pred.shape != data.y[:batch_size].shape
+            ):
+                raise Exception("Predictions have mismatched shape from provided y")
+            if isinstance(data.y, pd.Series) and (
+                mini_pred.shape != data.y.head(batch_size).shape
+            ):
+                raise Exception("Predictions have mismatched shape from provided y")
+        except Exception as e:
+            raise ValidationError(
+                "Lens.model predictions do not match expected form implied by provided labels y.",
+                e,
+            )
+
+    if "predict_proba" in model.__dict__.keys():
+        try:
+            mini_pred, batch_size = check_prediction_model_output(
+                model.predict_proba, data
+            )
+            if not mini_pred.size:
+                # results for all presently supported models are ndarray results
+                raise Exception("Empty return results from predict_proba function.")
+            if len(mini_pred.shape) > 1 and mini_pred.shape[1] > 1:
+                if np.sum(mini_pred[0]) != 1:
+                    raise Exception(
+                        "`predict_proba` outputs invalid. Per-sample outputs should sum to 1."
+                    )
+            else:
+                if mini_pred[0] >= 1:
+                    raise Exception(
+                        "`predict_proba` outputs invalid. Binary outputs should be <= 1."
+                    )
+        except Exception as e:
+            raise ValidationError(
+                "Lens.model outputs do not match expected form implied by provided labels y.",
+                e,
+            )
+
+    if "compare" in model.__dict__.keys():
```

Comment:
Do you think these validations could/should be defined within the model? It seems they are mostly dependent on the model functionality, with the data as input. What do you think of each model having a `validate_against_data` function that this lens_validation function calls?

Reply:
I don't really see it as being obviously better one way or the other.

This model + data validation step is only going to run once: when model + data are passed to the Lens object. I think there are equally strong arguments for that validation being a feature of the model (it's about its behavior on data) vs. Lens (we care about whether the things you passed for evaluation work together as a unit).

I tried to think of which way would lead to easier development down the road and I think the answer there is also "it's about equal". 
- In the paradigm where validation is a feature of the model, we'd make this existing code more modular (cut down on the if-statements in `check_model_data_consistency()`) and only call these helpers more places. We'd add helpers when there's new functionality like `my_funky_predict_function()`. If there's a new data type we'd have to go through all the existing models (or merely the helpers) to add some if statements (see e.g., lines 118-124 in the current version of this file).
- In the paradigm where validation lives primarily in lens (status quo in this PR), future development would mean adding ugly if-statements (that I think we need in scenario 1 anyway) to `check_model_data_consistency` and crafting a new helper for whatever functionality (for new models/prediction functions) or adding if-statements to helpers for new data types